flang/lib/Optimizer/Dialect/FIROps.cpp

```diff
@@ -346,10 +346,14 @@ static mlir::LogicalResult verify(fir::ArrayCoorOp op) {
       return op.emitOpError("number of indices do not match dim rank");
   }
 
-  if (auto sliceOp = op.slice())
+  if (auto sliceOp = op.slice()) {
+    if (auto sl = mlir::dyn_cast_or_null<fir::SliceOp>(sliceOp.getDefiningOp()))
+      if (!sl.substr().empty())
+        return op.emitOpError("array_coor cannot take a slice with substring");
```

Comment:
Substrings are objects and not proper values. It makes sense to allow them in boxes, which require objects and have their own type and have a "byte offset, byte width" implication built in. But it seems to unnecessarily complicate other value semantics. For instance, what is the type of `obj%arr(:)(i:j)`? Does it interfere with `obj%arr(:)(k:l)`? Since the base array is the same, do you really want to make 2 copies, one for each?